Added Readme and corresponding figure. (#47)
rasmusmarak/TOSS: README.md

# Trajectory Optimization of a Spacecraft Swarm (TOSS)

<!-- TABLE OF CONTENTS -->
<details open="open">
  <summary>Table of Contents</summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#installation">Installation</a></li>
        <li><a href="#test">Test</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#contributing">Contributing</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
  </ol>
</details>

<!-- ABOUT THE PROJECT -->

## About The Project

> Trajectory Optimization for a Spacecraft Swarm (TOSS) is a thesis project presented for the degree of Master of Science in mathematics, with specialization in Optimization and Systems theory provided by KTH Royal Institute of Technology. The thesis topic was defined within the scope of $\Phi$[-lab@Sweden](https://www.ai.se/en/data-factory/f-lab-sweden) in the frame of a collaboration between [AI Sweden](https://www.ai.se/en/) and the [European Space Agency](https://www.esa.int/) to explore distributed edge learning for space applications. The project was mainly supervised by Pablo Gómez and Emmanuel Blazquez from the Advanced Concepts Team (ACT) at ESA, and examined by Xiaoming Hu from KTH.  

Reviewer: I would move this into contact, most people won't care

Author: Fixed!

> The goal of the project is to enable an evaluation of the effects for single and multi-spacecraft architectures when applied to a prospective mission with specified science objectives. In particular, TOSS is designed to be:

Reviewer: Way too vague imo. You need to mention
- Optimizing trajectories with impulsive maneuvers around irregular bodies such as asteroids and comets using a polyhedral gravity model, RK xyz scheme and pygmo for evolutionary optimization.
- Goal of the trajectories is to cover the volume near the body well without getting too close

Author: Fixed!

- **open-source**: the source code of TOSS is publicly available.
- **decentralised**: each function and capability within the module can be adopted to generate trajectories for a wider variety of objectives and celestial bodies.
- **user-expandable**: the user can expand the capabilities of the code using the decentralised structure. 

> The optimization is based on an evolutionary approach sampling a large population where each candidate solution results in a deterministics trajectory that is obtained by numerical intergration. The trajectories are then resamples for a desired fixed time-step and then evaluating according to a defined fitness function. Then, unless the algorithm has met a certain stopping criteria, the population is mutated accoridng to the best candidate solution and the optimisation continuous by reiterating the process again.

Reviewer: ```suggestion
The optimization is based on an evolutionary approach sampling a large population where each candidate solution results in a deterministic trajectory that is obtained by numerical integration. The trajectories are then resamples for a desired fixed time-step and then evaluating according to a defined fitness function. Then, unless the algorithm has met a certain stopping criteria, the population is mutated according to the best candidate solution and the optimisation continuous by reiterating the process again.
```
would suggest to run a spellcheck on the file ;)

Author: Fixed!

<p align="center">
  <a href="https://github.com/rasmusmarak/TOSS">

>     <img src="docs/source/TOSS_Optimization_Scheme.pdf" alt="Scheme"  width="910" height="459">

Reviewer: doesn't render for me, can see online here https://github.com/rasmusmarak/TOSS/tree/Adding-Readme-to-repo

Author: Fixed for now, could perhaps be polished later on.

  </a>
  <p align="center">
    Trajectory Optimisation of a Spacecraft Swarm
    <br />
    <a href="https://github.com/rasmusmarak/TOSS/issues">Report Bug</a>
    ·
    <a href="https://github.com/rasmusmarak/TOSS/issues">Request Feature</a>
  </p>
</p>

### Built With

This project is based on:

- [polyhedral-gravity-model](https://github.com/esa/polyhedral-gravity-model) a fast, parallelised version of the polyhedral gravity model implemented in C++ along with a python interface.
- [DESolver](https://github.com/Microno95/desolver) a python library fro solving Initial Value Problems using numerical integrators. 

> - [PyGMo](https://esa.github.io/pygmo2/index.html) a scientific python library for massively parallel optimization managed by the Advanced Concepts Team at the European Space Agency

Reviewer: pygmo is C++ with a python wrapper I think

Author: Fixed!

For more details than provided by TOSS on these libraries, please refer to their docs.

<!-- GETTING STARTED -->

## Getting Started

This is a brief guide how to set up TOSS.

### Installation

### conda
The recommended way to install TOSS is via [conda](https://docs.conda.io/en/latest/) / [mamba](https://github.com/conda-forge/miniforge#mambaforge) using

```
conda install toss -c conda-forge

> ```

Reviewer: ?? I don't think we are on conda

Reviewer: if you want to go in that direction I would suggest we open an issue to add pip first as that is way easier but would not write either here until it is actually possible

Author: I removed this for now, but have added a new issue for this at #56.

### Building from source

Alternatively, to use the latest code from this repository make sure you have all the requirements installed and then clone the [GitHub](https://github.com/rasmusmarak/TOSS.git) repository as follows ([Git](https://git-scm.com/) required):.

```
git clone https://github.com/rasmusmarak/TOSS
```

To install TOSS you can use [conda](https://docs.conda.io/en/latest/) as follows:

```
cd TOSS
conda env create -f environment.yml
```

This will create a new conda environment called `toss` and install the required software packages.
To activate the new environment, you can use:

```
conda activate toss
```

### Test

After cloning the repository, developers can check the functionality of TOSS by running the following command in the `TOSS/tests` directory:

```sh
pytest
```

<!-- USAGE EXAMPLES -->

## Usage

### Config

> TOSS uses a central config file which is passed through the entire program. The default config parameters can be seen [here](https://github.com/rasmusmarak/TOSS/blob/2501351468eb362fa9ec8217b2fc426db37fca9a/toss/resources/default_cfg.toml). 

Reviewer: ```suggestion
TOSS uses a central config file which is passed through the entire program. The default config parameters can be seen [here](https://github.com/rasmusmarak/TOSS/blob/main/toss/resources/default_cfg.toml). 
```
Why link to a specific commit?

Author: Fixed!

> **[Coming Soon]** Practical usage of the config files will soon be added here through an explanatory notebook.

Reviewer: Maybe create an issue for it if you plan to do it? :)

Author: There is one in #48 and I plan to do this as soon as possible, but likely not until right after the thesis has been officially finished.

### Use Case 1: Trajectory Optimization of a Swarm Orbiting 67P/Churyumov-Gerasimenko Maximising Gravitational Signal

> In this case, the aim is to compute a set of trajectories corresponding to a spacecraft swarm of n spacecraft, each with m impulsive maneuvers. The simulation is done within the context of a prospective mission around the comet 67P/Churyumov-Gerasimenko and the scientific target of maximising the measured gravitational signal. For details on mission parameters, optimization structure and fitness functions, see the corresponding paper on [ArXiv](https://doi.org/10.48550/arXiv.2306.01602).

Reviewer: I would suggest to use this https://citation-file-format.github.io/ , would also mention the paper higher up somewhere

Author: Added the cff-file in PR #54. Maybe we could merge that branch into this one for simplicity, and then I'll change the referencing in this PR directly before merging to main?

Reviewer: We can but it seems you branched out of main in #54 , thus it would be easier to merge them individually I think

To run this test case, install TOSS and setup the environment as presented above, then run the following:
```
OMP_NUM_THREADS=1 python toss.py
```

> TOSS will then return the four csv files

Reviewer: ```suggestion
TOSS will then create four csv files
```

Author: Fixed!

- **run_time**: which refers to the total simulation time.
- **champion_f**: being the fitness value corresponding to the champion chromosome.
- **champion_x**: being the champion chromosome corresponding to the optimal set of trajectories.
- **fitness_list**: which is a list of champion fitness values correponding to each generation.

### Use Case 2: Trajectory Optimization Around an Arbitrary Clestial Body Maximising Gravitation Signal
**[Coming Soon]**

### Use Case 3: Trajectory Optimization with User-Defined Scientific Objective

> **[Coming Soon]**

Reviewer: Issue for it?

Author: Added a new issue for this at #55.

2. Get the most up-to-date code by following this quick guide for installing TOSS from source:
   1. Get [miniconda](https://docs.conda.io/en/latest/miniconda.html) or similar
   2. Clone the repo
   ```sh
   git clone https://github.com/rasmusmarak/TOSS
   ```
   3. Set up the environment. This creates a conda environment called
      `toss` and installs the required dependencies.
   ```sh
   conda env create -f environment.yml
   conda activate toss
   ```

Once the installation is done, you are ready to contribute.
Please note that Pull-Requests should be created from and into the `main` branch.

3. Create your Feature Branch (`git checkout -b feature/yourFeature`)
4. Commit your Changes (`git commit -m 'Add some yourFeature'`)
5. Push to the Branch (`git push origin feature/yourFeature`)
6. Open a Pull Request on the `main` branch.

and we will have a look at your contribution as soon as we can.

Furthermore, please make sure that your Pull-Request passes all automated tests. Review will only happen after that.
Only Pull-Requests created on the `main` branch with all tests passing will be considered.

<!-- LICENSE -->

## License

> **[Coming Soon]** Distributed under the .... License. See [LICENSE]() for more information.

Reviewer: need to update

Author: Referenced in issue #57 and added in PR #58.
